modules/margin-protocol/src/lib.rs

```diff
@@ -341,28 +339,29 @@ impl<T: Trait> Module<T> {
 			.ok_or(Error::<T>::NumOutOfBound)?;
 
 		// realizing
-		let balance_delta_abs = u128_from_fixed_128(balance_delta.saturating_abs());
 		if balance_delta.is_positive() {
-			// trader has profit
-			let realized = cmp::min(
-				<T::LiquidityPools as LiquidityPools<T::AccountId>>::liquidity(position.pool),
-				balance_delta_abs,
-			);
+			// trader has profit, max realizable is the pool's liquidity
+			let pool_liquidity = fixed_128_from_u128(<T::LiquidityPools as LiquidityPools<T::AccountId>>::liquidity(
+				position.pool,
+			));
+			let realized = cmp::min(pool_liquidity, balance_delta);
 			<T::LiquidityPools as LiquidityPools<T::AccountId>>::withdraw_liquidity(
 				&Self::account_id(),
 				position.pool,
-				realized,
+				u128_from_fixed_128(realized),
 			)?;
-			<Balances<T>>::mutate(who, |b| *b += realized);
+			Self::_update_balance(who, realized);
 		} else {
-			// trader has loss
-			let realized = cmp::min(Self::balances(who), balance_delta_abs);
+			// trader has loss, max realizable is the trader's equity
+			let equity = Self::_equity_of_trader(who)?;
+			let balance_delta_abs = balance_delta.saturating_abs();
+			let realized = cmp::min(equity.saturating_add(balance_delta_abs), balance_delta_abs);
 			<T::LiquidityPools as LiquidityPools<T::AccountId>>::deposit_liquidity(
```

Comment:
What happens if the margin protocol doesn't have enough funds to pay the balance_delta amount to the liquidity pool?

Reply:
Closing position would fail and return an error then.

Reply:
But this should never happen in theory, right?

Reply:
Technically it would happen if:
- Risk parameters not set properly, and
- Market fluctuating massively, and
- Liquidation got delayed, and
- Tons of loss in general and protocol's lock account balance run dry.

Reply:
So then better to send `max(margin_protocol_balance, balance_delta)`? As safety and then the trader can at least close his position.

Reply:
As discussed in the channel, for now we'll not make margin protocol to cover the loss. Later there will be compensation design for this kind of extreme scenario, but out of the scope of this PR.